internal/rukpak/preflights/crdupgradesafety/checks.go

```diff
@@ -70,3 +73,246 @@ func (c *ServedVersionValidator) Validate(old, new apiextensionsv1.CustomResourc
 func (c *ServedVersionValidator) Name() string {
 	return "ServedVersionValidator"
 }
+
+type resetFunc func(diff kappcus.FieldDiff) kappcus.FieldDiff
+
+func isHandled(diff kappcus.FieldDiff, reset resetFunc) bool {
+	diff = reset(diff)
+	return reflect.DeepEqual(diff.Old, diff.New)
+}
+
+func Enum(diff kappcus.FieldDiff) (bool, error) {
+	reset := func(diff kappcus.FieldDiff) kappcus.FieldDiff {
+		diff.Old.Enum = []apiextensionsv1.JSON{}
+		diff.New.Enum = []apiextensionsv1.JSON{}
+		return diff
+	}
+
+	oldEnums := sets.New[string]()
+	for _, json := range diff.Old.Enum {
+		oldEnums.Insert(string(json.Raw))
+	}
+
+	newEnums := sets.New[string]()
+	for _, json := range diff.New.Enum {
+		newEnums.Insert(string(json.Raw))
+	}
+	diffEnums := oldEnums.Difference(newEnums)
+	var err error
+
+	switch {
+	case oldEnums.Len() == 0 && newEnums.Len() > 0:
+		err = fmt.Errorf("enum constraints %v added when there were no restrictions previously", newEnums.UnsortedList())
+	case diffEnums.Len() > 0:
+		err = fmt.Errorf("enums %v removed from the set of previously allowed values", diffEnums.UnsortedList())
+	}
+
+	return isHandled(diff, reset), err
+}
+
+func Required(diff kappcus.FieldDiff) (bool, error) {
+	reset := func(diff kappcus.FieldDiff) kappcus.FieldDiff {
+		diff.Old.Required = []string{}
+		diff.New.Required = []string{}
+		return diff
+	}
+
+	oldRequired := sets.New(diff.Old.Required...)
+	newRequired := sets.New(diff.New.Required...)
+	diffRequired := newRequired.Difference(oldRequired)
+	var err error
+
+	if diffRequired.Len() > 0 {
+		err = fmt.Errorf("new required fields %v added", diffRequired.UnsortedList())
+	}
+
+	return isHandled(diff, reset), err
+}
+
+func Maximum(diff kappcus.FieldDiff) (bool, error) {
+	reset := func(diff kappcus.FieldDiff) kappcus.FieldDiff {
+		diff.Old.Maximum = nil
+		diff.New.Maximum = nil
+		return diff
+	}
+
+	var err error
+
+	switch {
+	case diff.Old.Maximum == nil && diff.New.Maximum != nil:
+		err = fmt.Errorf("maximum constraint %v added when there were no restrictions previously", *diff.New.Maximum)
+	case diff.Old.Maximum != nil && diff.New.Maximum != nil && *diff.New.Maximum < *diff.Old.Maximum:
+		err = fmt.Errorf("maximum constraint decreased from %v to %v", *diff.Old.Maximum, *diff.New.Maximum)
+	}
```

Comment:
This switch statement is for the most part repeated across all the `Max*` functions. I'm planning to make an update to have a function that is re-used across all of them, but didn't have time left and wanted to get this PR up before I logged off for the day. I figure it's not the end of the world if this goes in as-is since I could file a follow-up. If it hasn't merged by the time I get to circle back around to it I'll push the update :)

Reply:
Done in https://github.com/operator-framework/operator-controller/pull/1269/commits/787b14815e628f68fae683e2125f6b4289dd380e